---
title: "The dbfmcl library"
author: "J. Bavais, Aurelie Bergon, Fabrice Lopez, Julien Textoris, L. Spinelli and D. Puthier"
date: "`r Sys.Date()`"
abstract: >
  A basic task in the analysis of transcriptomics data is the selection of features (genes) and their partitioning to obtain a set of gene clusters that are expected to highlight biological functions, pathways, cell heterogeneity (...). Performing clustering automatically is a daunting task as the number of underlying gene clusters is most generally unknown before starting the analysis. The dbfmcl package is intended to propose a set of functions to infer, in an automated fashion a reasonnable solution to provide, without too much parametrization a first overview of the dataset.
  dbfmcl package version: `r packageVersion("dbfmcl")`
output:
  rmarkdown::html_document:
    highlight: pygments
    toc: true
    fig_width: 5
bibliography: library.bib
vignette: >
  %\VignetteIndexEntry{Analyzing scRNA-seq data with DBFMCL}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
  %\usepackage[utf8]{inputenc}
---

```{r setup, echo=FALSE, results="hide"}
knitr::opts_chunk$set(tidy = FALSE,
                      cache = FALSE,
                      dev = "png",
                      message = FALSE, error = FALSE, warning = TRUE)
```


# Introduction

**Note:** if you use dbfmcl in published research, please cite:

TODO

One important issue in RNA-seq and scRNA-seq is the selection of informative features (i.e genes) and the identification of gene clusters. The scRNA-seq datasets are considered to be highly noisy and a typical way of selecting features is to filter them based on variance. Regarding gene partitioning a popular solution is to use k-means. In the context of the microarray data analysis we had shown that another way of selecting informative genes is to try to capture groups of highly correlated features (clusters). These genes can be viewed as elements displaying weak distances to their nearest neighbors that represent densely populated regions  (i.e. strong profile similarities). This solution was previously implemented in the DBFMCL (Density-Based Filtering and Markov Clustering) algorithm. To isolate these regions DBF-MCL computes, for each gene/element, the distance with its kth nearest neighbor (DKNN). In order to define a critical DKNN value (that will depend on the dataset), DBF-MCL computes simulated DKNN values by using an empirical randomization procedure. Given a dataset containing n genes and p samples, a simulated DKNN value is obtained by sampling n distance values from the gene-gene distance matrix D and by extracting the kth-smallest value. This procedure is repeated n times to obtain a set of simulated DKNN values S. Computed distributions of simulated DKNN are used to
compute a FDR value for each observed DKNN value. The critical value of DKNN is the one for which a user-defined FDR value (typically 10%) is observed. Genes with DKNN value below this threshold are selected and used to construct a graph. In this graph, edges are constructed between two genes (nodes) if one of them belongs to the k-nearest neighbors of the other. Edges are weighted based on
the respective coefficient of correlation (i.e., similarity) and the graph obtained is partitioned using the Markov CLustering algorithm (MCL).
While this algorithm was previously developed for microarray we have adpated it to work with discrete data, especially scRNA-seq data.

# Applying DBFMCL to PBMC dataset

For this tutorial, we will use the classical Seurat dataset related to Peripheral Blood Mononuclear Cells (PBMC) that is freely available from 10X Genomics. This [dataset](https://cf.10xgenomics.com/samples/cell/pbmc3k/pbmc3k_filtered_gene_bc_matrices.tar.gz) contains 2,700 single cells sequenced on the Illumina NextSeq 500. 
  
  
```{r}
# This is for developement purpose
path_to_dbfmcl_git = "/Users/puthier/Documents/git/project_dev/dbfmcl/"
reinstall <- function(){
  # For development purpose
  # please adapt the path
  try(detach("package:dbfmcl"), silent = TRUE)
  library(devtools)
  roxygen2::roxygenise("/Users/puthier/Documents/git/project_dev/dbfmcl/")
  install(path_to_dbfmcl_git)
}
reinstall()
library(dbfmcl)
```

```{r}
library(Seurat)
library(dbfmcl)

# Load the PBMC dataset
# please adapt the path
setwd("~/Downloads/")
pbmc_data <- Read10X(data.dir = "filtered_gene_bc_matrices/hg19/")
pbmc_raw <- CreateSeuratObject(counts = pbmc_data, project = "pbmc3k",
                               min.cells = 3, 
                               min.features = 200)

# Perform simple log-transformation and scaling
pbmc_norm <- NormalizeData(pbmc_raw, 
                          normalization.method = "LogNormalize",
                          scale.factor = 10000)


```

From this Seurat object we can directly perform feature selection and gene cluster inference. Here *DBFMCL()* is used with default arguments.

```{r}
pbmc_df <- as.data.frame(pbmc_norm@assays$RNA@data)
dbf <- DBFMCL(pbmc_df)
```

Note that the result of the DBFMCL function is an S4 object with the following slots:

```{r}
slotNames(dbf)
```

# Search of cell populations using DBFMCL selected genes

The gene filtered using DBFMCL can now be used to subset a Seurat Object and search for cell populations.

```{r}
pbmc_dbf <- subset(pbmc_norm,  features=rownames(dbf@data))
pbmc_dbf <- ScaleData(pbmc_dbf)
pbmc_dbf <- RunPCA(pbmc_dbf, features = rownames(pbmc_dbf))
pbmc_dbf <- RunUMAP(pbmc_dbf, dims = 1:10)
pbmc_dbf <- FindNeighbors(pbmc_dbf, reduction = "pca", dims = 1:20)
pbmc_dbf <- FindClusters(pbmc_dbf)
dim_plot_clust <- DimPlot(pbmc_dbf, reduction = "umap")
dim_plot_clust
```

# Transfer infered cell populations to a DBFMCL object

The inferred populations can now be transferred into the ClusterSet object and represented using the plot_clust function.

```{r}
dbf_seurat <- load_seurat(dbf, 
                          seurat_obj=pbmc_dbf,
                          dimplot_obj=dim_plot_clust)
g_profiles <- plot_clust(dbf_seurat,
       average_only = TRUE,
       floor=-1, 
       ceil=1)
g_profiles
```

Both the dimplot and gene profiles can be presented in the same diagram using the command below

```{r}
g_profiles + dim_plot_clust
```



